test/with_dummyserver/test_socketlevel.py

```diff
@@ -1939,3 +1941,163 @@ def socket_handler(listener: socket.socket) -> None:
+        self, method: str, content_length: Optional[int]
+    ) -> None:
+        buffer = bytearray()
+
+        def socket_handler(listener: socket.socket) -> None:
+            nonlocal buffer
+            sock = listener.accept()[0]
+            while not buffer.endswith(b"\r\n\r\n"):
+                buffer += sock.recv(65536)
+            sock.sendall(
+                b"HTTP/1.1 200 OK\r\n"
+                b"Server: example.com\r\n"
+                b"Content-Length: 0\r\n\r\n"
+            )
+            sock.close()
+
+        self._start_server(socket_handler)
+
+        headers = {}
+        if content_length is not None:
+            headers["Content-Length"] = str(content_length)
+
+        with HTTPConnectionPool(self.host, self.port, timeout=3) as pool:
+            resp = pool.request(method, "/")
+            assert resp.status == 200
+
+        sent_bytes = bytes(buffer)
+        assert b"Accept-Encoding: identity\r\n" in sent_bytes
+        assert b"Content-Length: 0\r\n" in sent_bytes
+        assert b"transfer-encoding" not in sent_bytes.lower()
+
+    @pytest.mark.parametrize("chunked", [True, False])
+    @pytest.mark.parametrize("method", ["POST", "PUT", "PATCH"])
+    @pytest.mark.parametrize("body_type", ["file", "generator", "bytes"])
+    def test_chunked_specified(
+        self, method: str, chunked: bool, body_type: str
+    ) -> None:
+        buffer = bytearray()
+
+        def socket_handler(listener: socket.socket) -> None:
+            nonlocal buffer
+            sock = listener.accept()[0]
+            sock.settimeout(0)
+
+            start = time.time()
+            while time.time() - start < 0.5:
+                try:
+                    buffer += sock.recv(65536)
+                except OSError:
+                    continue
+
+            sock.sendall(
+                b"HTTP/1.1 200 OK\r\n"
+                b"Server: example.com\r\n"
+                b"Content-Length: 0\r\n\r\n"
+            )
+            sock.close()
+
+        self._start_server(socket_handler)
+
+        body: Any
+        if body_type == "generator":
+
+            def body_generator() -> Generator[bytes, None, None]:
+                yield b"x" * 10
+
+            body = body_generator()
+        elif body_type == "file":
+            body = io.BytesIO(b"x" * 10)
+            body.seek(0, 0)
+        else:
+            if chunked is False:
+                pytest.skip("urllib3 uses Content-Length in this case")
+            body = b"x" * 10
+
+        with HTTPConnectionPool(self.host, self.port, timeout=3, retries=False) as pool:
+            resp = pool.request(method, "/", chunked=chunked, body=body)
+            assert resp.status == 200
+
+        sent_bytes = bytes(buffer)
+        assert sent_bytes.count(b":") == 5
+        assert b"Host: localhost:" in sent_bytes
+        assert b"Accept-Encoding: identity\r\n" in sent_bytes
+        assert b"Transfer-Encoding: chunked\r\n" in sent_bytes
```

Comment:
Why is this true when `chunked` is false?

Reply:
`chunked=True` essentially means "force chunked" and `chunked=False` (for both urllib3 and stdlib) means use `content-length` if you can find the length of `body` otherwise use `transfer-encoding: chunked`.